### DISCUSSION TITLE
Enhancement: Choose what messages from context are being sent

### DISCUSSION THREAD

**Discussion by Trapper4888** [danny-avila/LibreChat#2377] [Feature Requests & Suggestions]
### What features would you like to see added?

Implement a feature that allows users to customize the number of messages sent as context with each new message. This feature will help manage costs and provide users with fine-grained control over the conversation context.

### More details

1. Add an input field, maybe next to the preset button, allowing users to select the number of messages to include in the context window. If no sliding window is specified, all messages in the conversation history will be sent as context, excluding any "force-discarded" messages.

2. Introduce "force-ping" and "force-discard" options for individual messages in the conversation history.
    - "Force-ping" ensures that a specific message is always included in the context, regardless of its position or the sliding window settings.
    - "Force-discard" excludes a message from the context, regardless of the sliding window settings.

3. Default behaviours :
    - When no sliding window is specified, and no messages are "force-discarded," the entire conversation history is sent as context.
    - When a sliding window is specified, and no messages are "force-pinned" or "force-discarded," the context window includes the last 'x' messages from the conversation, where 'x' is the number selected by the user.

4. "Force-ping" and "force-discard" behaviour :

    - If a message is "force-pinned," it is always included in the context, along with the 'x' last message (when a sliding window is specified).
    - If a message is "force-discarded," it is always excluded from the context, regardless of the sliding window settings.

### Which components are impacted by your request?

General, UI

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by danny-avila** [comment]
Better managing conversation history on the message level is something I'm actively working on. First by tokens, then by messages.

About "pinning" messages. It can be unreliable as forcing a message can throw an error if it exceeds the context window, or you compensate too much to include it (in the extreme case, it's the only message included, and/or it throws an API error).

**Comment by Trapper4888** [comment]
I see, thank you for the reply.

This pinning feature could also be apart from regular messages:
We could have a context (that is not system, since not all models have system, and system may be used for another purpose) box, where we could put this text (like a code snippet).
Then, LibreChat would format it and send it as an extra first user message in a best effort manner: if it can fit, let's fit it in, if it can't, it can't. But if it takes less than let's say 20% of the max token to send, and there are at least 3-4 messages that can fit after it, it will be included in a request even if the first messages of the conversation are discarded.
There could even be a small button on messages "send-to-context" that would append a message to this context box.

We could even have an optional "summary of context" that would work like the response titling: a model to summarize this context if it needs to be squeezed in a shorter window. In a last effort, the context could be passed by parts like a RAG document.

This could also resolve the per-message/conversation-wide pdf document. 

___

### Here is a Claude 3 clearer formulation

- Introducing a dedicated "context box" separate from regular messages, where users can input text or code snippets to provide context for the conversation. (!= system message)

- LibreChat would format this context and send it as an extra first user message in a best-effort manner:
    - If the context fits within the message limit, include it.
    - If the context doesn't fit, but takes up less than 20% of the max token limit and allows at least 3-4 messages to follow, include it even if the first messages of the conversation are discarded.

- Implementing a "send-to-context" button on messages, allowing users to append a message to the context box easily.

- Optionally, introducing a "summary of context" feature, similar to response titling, where a model summarizes the context if it needs to be condensed to fit within a shorter window.

- As a last resort, the context could be passed in parts, similar to a RAG (Retrieval-Augmented Generation) document.

- This feature could also address the issue of handling per-message or conversation-wide PDF documents by allowing to include documents in the context.

**Comment by sweitzel** [comment]
I believe its rather difficult of solving the the conversation memory issue efficiently. In my opinion this should be simple, intuitive for the user, and work in the background without requiring additional user intervention.
There are many different concept like embedding the context, context compression, step-back prompting, etc. and I find it hard to chose the best approach.

As a first step it would be quite useful to just give an indicator to the user of which messages are included in the conversation context and which not. The messages which are not included in the context could be grayed out for example. 
Of course all this just makes sense after the conversation history limit feature mentioned [here ](https://github.com/danny-avila/LibreChat/discussions/2308#discussioncomment-9009244) is implemented.

**Comment by leiyangyou** [comment]
I think we can embed the conversation as we go, and use embeddings to search for relevant messages first, then append the last N as context window limit allows
